### PR TITLE
fix #1127,  replace spray with akka-http..

### LINKED ISSUE #1127: Master shuts down with out of memory
[Story](https://www.pivotaltracker.com/story/show/98513158)
in kafka-hdfs-pipeline example master will exit the JVM. Log out follows...

### PR DESCRIPTION
**Pull request by clockfly**

### CODE REVIEW

**Comment by clockfly** [comment]
not ready for merge yet.

File: daemon/src/main/scala/io/gearpump/util/FileDirective.scala [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        }
+      }
+    }
+  }
+
+  def downloadFile(file: File): Route = {
+    val responseEntity = HttpEntity(
+      MediaTypes.`application/octet-stream`,
+      file.length,
+      SynchronousFileSource(file, chunkSize = 262144))
```
Comment (stanleyxu2005) [inline daemon/src/main/scala/io/gearpump/util/FileDirective.scala]: Where the magic number `262144` comes from? Is it not configurable?

File: services/jvm/src/main/scala/io/gearpump/services/MasterService.scala [hunk 2]
```diff
@@ -33,194 +36,185 @@ import io.gearpump.cluster.worker.WorkerSummary
...
+                uploadFile { fileMap =>
+                  val jar = fileMap.get("jar").map(_.file)
+                  val userConf = fileMap.get("conf").map(_.file)
+
+                  if (jar.isEmpty) {
                     complete(write(
-                      MasterService.Status(success=true)))
-                  case Failure(ex) =>
-                    failWith(ex)
+                      MasterService.Status(success = false, reason = "Jar file not supplied")))
```
Comment (stanleyxu2005) [inline services/jvm/src/main/scala/io/gearpump/services/MasterService.scala]: I'd recommend `MasterService.Status` to be `MasterService.Response` or `RestService.Response`.

File: daemon/src/test/scala/io/gearpump/util/FileServerSpec.scala [hunk 3]
```diff
@@ -78,14 +85,12 @@ class FileServerSpec  extends WordSpecLike with Matchers with BeforeAndAfterAll
...
-
-      val url = s"http://$host:${port.port}/notexist"
+      val server = new FileServer(system, host, 0, rootDir)
+      val port = Await.result((server.start), Duration(25, TimeUnit.SECONDS))
 
-      val client = FileServer.newClient
-      val fetchedBytes = client.get(url)
-      assert(fetchedBytes.isFailure, s"fetch data fails, $url, $rootDir")
+      val client = new Client(system, host, port.port)
+      val fetchedBytes = get(client, FilePath("noexist"))
```
Comment (stanleyxu2005) [inline daemon/src/test/scala/io/gearpump/util/FileServerSpec.scala]: `FilePath("noexist")` might hit a physical file accidentally

File: daemon/src/test/scala/io/gearpump/util/FileServerSpec.scala [hunk 4]
```diff
@@ -78,14 +85,12 @@ class FileServerSpec  extends WordSpecLike with Matchers with BeforeAndAfterAll
...
+      println(fetchedBytes)
```
Comment (stanleyxu2005) [inline daemon/src/test/scala/io/gearpump/util/FileServerSpec.scala]: Remove the debug code. Unittest does not need a `println`

File: daemon/src/main/scala/io/gearpump/util/FileServer.scala [hunk 5]
```diff
@@ -1,137 +1,125 @@
...
-      val data = status.flatMap { code =>
-        if (code == 200) {
-          Success(get.getResponseBody())
-        } else {
-          Failure(new Exception(s"We cannot get the data, the status code is $code"))
-        }
-      }
-      data
+    private def entity(file: File)(implicit ec: ExecutionContext): Future[RequestEntity] = {
+      val entity =  HttpEntity(MediaTypes.`application/octet-stream`, file.length(), SynchronousFileSource(file, chunkSize = 100000))
```
Comment (stanleyxu2005) [inline daemon/src/main/scala/io/gearpump/util/FileServer.scala]: `chunkSize` should not be hard coded.

**Pull request closed without merging**